docs/long_term_roadmap.md

```diff
@@ -0,0 +1,221 @@
+is not in the business of "innovating" either on the frontend or backend sides.
+The project scope is to be an enabling connector between the two systems.
+
+Non-technically, Torch-MLIR's goal is not to be an end-to-end product, but a
+reliable piece of "off the shelf" infrastructure that system designers use as
+part of their larger end-to-end flows. The main users are expected to be
+"integrators", not end-users writing Python. This has the following facets:
+- Community: Users of Torch-MLIR should feel empowered to participate in the
+  community to get their questions resolved, or propose (and even implement)
+  changes needed for their use cases.
+- Ecosystem alignment: Users of Torch-MLIR should feel that the project is
+  aligned with all of the projects that it collaborates with, making it safe to
+  bet on for the long term.
+- Ease of use: Users of Torch-MLIR should feel that it "Just Works", or that
+  when it fails, it fails in a way that is easy to understand, debug, and fix.
+- Development: Torch-MLIR should be easy and convenient to develop.
+
+Today, much of the design space and the main problems have been identified, but
+larger-scale architectural and cross-project changes are needed to realize the
+right long-term design. This will allow us to reach a steady-state that best
+meets the goals above.
+
+## The main architectural changes
+
+As described in
+[architecture.md](architecture.md),
+Torch-MLIR can be split into two main parts: the "frontend" and the "backend".
+
+The main sources of brittleness, maintenance cost, and duplicated work across
+the ecosystem are:
+
+- The frontend work required to lower TorchScript to the backend contract.
+- The irregular support surface area of the large number of PyTorch ops across
+  the Linalg, TOSA, and MHLO backends.
+
+Most of this document describes long-term ecosystem changes that will address
+these, drastically improving Torch-MLIR's ability to meet its goals.
+
+## Roadmap
+
+### Refactoring the frontend
+
+The primary way to make the frontend more reliable is to leverage new PyTorch
+infrastructure that bridges from the PyTorch eager world into compiler-land.
+PyTorch has two main projects that together cover almost all user use cases and
+provide a technically sound, high quality-of-implementation path from user
+programs into the compiler.
+
+- [TorchDynamo](https://github.com/pytorch/torchdynamo) - TorchDynamo uses
+  tracing-JIT-like techniques and program slicing to extract traces of tensor
+  operations, which can then be passed to lower-level compilers. It works
+  seamlessly with unmodified user programs.
+- [FuncTorch](https://github.com/pytorch/functorch) - FuncTorch is basically JAX
+  for PyTorch. It requires manual program tracing and slicing, but that is
+  actually important for users since it gives them direct control over various
+  important transformations, such as `grad` and `vmap`.
```

Comment:
Does this mean that LTC support will eventually be deprecated in Torch-MLIR?

Reply:
As long as LTC is not deprecated by PyTorch and we have users that rely on it, then I think we should keep supporting it.

But as Will said [here](https://dev-discuss.pytorch.org/t/skipping-dispatcher-with-lazytensor/634/2?u=_sean_silva), it seems like on the 1-2yr time horizon LTC will be more an implementation detail under TorchDynamo for users that already have compilers written using LTC. That is, LTC is basically just a way to convert a TorchDynamo FX graph into LTC graphs, for users that have toolchains written against LTC graphs. But that won't make much sense for us, because we convert to MLIR in the end no matter what. That is, in the future going `TorchDynamo FX graph -> LTC Graph -> MLIR` can just be replaced by the direct `TorchDynamo FX graph -> MLIR` path. So I predict that in the 1-2yr time horizon we will be able to delete our LTC support.

The only blocker there would be if end-users have `device='lazy'` hardcoded into their code. That will require a migration plan for current LTC-based toolchains onto TorchDynamo. This migration will improve the end-user experience since TorchDynamo is more seamless, but it is a end-user-impacting migration nonetheless and we will want to phase it appropriately with the community. Do you have a sense of how difficult such a migration would be for you folks?

Reply:
I added this to the doc.

Reply:
> That is, in the future going TorchDynamo FX graph -> LTC Graph -> MLIR can just be replaced by the direct TorchDynamo FX graph -> MLIR path. So I predict that in the 1-2yr time horizon we will be able to delete our LTC support.

If we can get the MLIR graph without any decompositions, then we can probably make this work.

> Do you have a sense of how difficult such a migration would be for you folks?

Having said that, our infrastructure is pretty heavily reliant on the LTC structure. It will take time for us to experiment with what it might take to be able to use TorchDynamo as our frontend instead. For that reason, I think keeping the `TorchDynamo FX graph -> LTC Graph -> MLIR` path would be very useful until such a time

Reply:
Yes, it will require design work for sure, and we will want to keep the lights on until the migration can happen.